relay-metrics/src/bucket.rs

```diff
@@ -648,6 +799,50 @@ impl MetricsContainer for Bucket {
     }
 }
 
+/// Iterator over parsed metrics returned from [`Metric::parse_all`].
+#[derive(Clone, Debug)]
+pub struct ParseBuckets<'a> {
+    slice: &'a [u8],
+    timestamp: UnixTimestamp,
+}
+
+impl Default for ParseBuckets<'_> {
+    fn default() -> Self {
+        Self {
+            slice: &[],
+            // The timestamp will never be returned.
+            timestamp: UnixTimestamp::from_secs(4711),
+        }
+    }
+}
+
+impl Iterator for ParseBuckets<'_> {
+    type Item = Result<Bucket, ParseBucketError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.slice.is_empty() {
+                return None;
+            }
+
+            let mut split = self.slice.splitn(2, |&b| b == b'\n');
+            let current = split.next()?;
+            self.slice = split.next().unwrap_or_default();
+
+            let string = match std::str::from_utf8(current) {
+                Ok(string) => string.strip_suffix('\r').unwrap_or(string),
+                Err(_) => return Some(Err(ParseBucketError(()))),
+            };
+
+            if !string.is_empty() {
```

Comment:
If this branch isn't matched, you go to the end of the loop, and then it goes into the next iteration. In this specific case, it will continue to consume lines from the slice until the slice is empty and then it goes to the explicit `return None`.